crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2589,6 +2640,25 @@ impl<'db> TypeInferenceBuilder<'db> {
                 }
             }
 
+            (
+                Type::BooleanLiteral(b1),
+                Type::BooleanLiteral(b2),
+                ruff_python_ast::Operator::BitOr,
+            ) => match (b1, b2) {
+                (true, true) => Type::BooleanLiteral(true),
+                (true, false) => Type::BooleanLiteral(true),
+                (false, true) => Type::BooleanLiteral(true),
+                (false, false) => Type::BooleanLiteral(false),
+            },
+
+            (Type::BooleanLiteral(_), right, op) => {
+                let int_value = left_ty.expect_int_literal();
+                self.infer_binary_expression_type(expr, op, Type::IntLiteral(int_value), right)
```

Comment:
Let's do it this way instead, like it's already done in `infer_unary_expression`:
```suggestion
            (Type::BooleanLiteral(bool_value), right, op) => {
                self.infer_binary_expression_type(expr, op, Type::IntLiteral(i64::from(bool_value)), right)
```
And same for the case below.

Bool and int literals aren't equivalent; the conversion between them should be explicit, it shouldn't be done implicitly by `expect_int_literal`.